File: tests/typesafety/python/config_validate_everywhere.py
# expect: passes
# title: The same check, repeated by every reader
"""Settings travel as a dict of strings. Every reader re-checks the port,
because no reader can trust that another already did."""


def read_argv(argv: list[str]) -> dict[str, str]:
    settings: dict[str, str] = {}
    for pair in argv:
        key, _, value = pair.partition("=")
        settings[key] = value
    return settings


def connect(settings: dict[str, str]) -> str:
    port_text = settings.get("port")
    if port_text is None or not port_text.isdigit():
        raise ValueError("port missing or not a number")
    return f"connecting on {int(port_text)}"


def report(settings: dict[str, str]) -> str:
    port_text = settings.get("port")
    if port_text is None or not port_text.isdigit():  # the same check, again
        raise ValueError("port missing or not a number")
    return f"listening on {int(port_text)}"

File: tests/typesafety/typescript/config_validate_everywhere.ts
// expect: passes

function readArgv(argv: string[]): Map<string, string> {
  const settings = new Map<string, string>();
  for (const pair of argv) {
    const [key, value] = pair.split("=", 2);
    settings.set(key ?? "", value ?? "");
  }
  return settings;
}

export function connect(settings: Map<string, string>): string {
  const portText = settings.get("port");
  if (portText === undefined || !/^\d+$/.test(portText)) {
    throw new Error("port missing or not a number");
  }
  return `connecting on ${Number(portText)}`;
}

export function report(settings: Map<string, string>): string {
  const portText = settings.get("port");
  if (portText === undefined || !/^\d+$/.test(portText)) { // the same check, again
    throw new Error("port missing or not a number");
  }
  return `listening on ${Number(portText)}`;
}
